### PR TITLE
VACMS-18821: implement script to move unattached pdfs from drupal into s3 bucket

### LINKED ISSUE #18821: Implement Script to move unattached PDFs from Drupal into S3 Bucket
## User Story or Problem Statement
Unattached PDFs associated with VHA Digital Media need to be removed from Drupal CMS and inserted into an S3 bucket. 

## Description or Additional Context 
Right now, there are unattached PDFs within Drupal CMS that need to be moved out of Drupal and into a S3 bucket. An unattached PDF is a PDF that is not connect to any content within Drupal CMS. The CMS team has been working with VHA Digital Media and have agreed to the idea of creating a script to clear out any unused PDFs. We also need the PDFs that are moved into the S3 bucket to be easily searchable, in case an editor needs the PDF or for a FOIA request.

## Steps for Implementation
- Create a script to identify the unattached PDFs
- Create a script to remove unattached PDFs from Drupal and into a S3 bucket
- Generate a log file of the unattached PDFs that were removed from Drupal and include metadata about the PDFs (e.g. File Name, Upload Date, Uploaded By). 
- Limit the access to the S3 bucket to only CMS engineers 

## Anti-Goals
- Do not run script until we've received a sign off from VHA Digital Media
- A user cannot search for specific PDFs within the S3 bucket. 

## Acceptance Criteria
- [x] Unattached PDFs associated with VHA digital media are ready to be removed from Drupal CMS and inserted in an S3 bucket.
- [x] A log file can be generated of all unattached PDFs removed from Drupal that includes the following metadata: File Name, Upload Date, and/or Upload By (username) in .csv format and inserted into the S3 bucket. 
- [x] The unattached PDFs are searchable by Files Name in the S3 bucket.

### PR DESCRIPTION
**Pull request by edmund-dunn**
## Description

Closes #18821 

The env vars have been set up in the devops repo for `cms` and `cms-test`

## Testing done
Tested locally

## QA steps

What needs to be checked to prove this works?
What needs to be checked to prove it didn't break any related things?
What variations of circumstances (users, actions, values) need to be checked?

1. Go to the tugboat dashboard and open terminal for `php` server in tugboat dashboard and run `drush php-eval "print_r(\Drupal::service('va_gov_media.pdf_delete')->vamcPdfDelete());"`
   - [x] Validate that the script runs without errors.
2. Then go the `/admin/reports/dblog`
   - [x] Validate that you see a "CSV written to ..." message
   - [x] Validate that you see a "PDFs deleted and CSV written" message
   - [x] Validate that you see a "Total PDFs: " message
   - [x] Validate that you see a "VHA unattached PDFs deleted: " message
   - [x] Validate that you see a "Non-VHA unattached PDFs not deleted: " message
   - [x] Check for errors in the `va_gov_media` channel
   - [x] Check for warnings in the `va_gov_media` channel

### Definition of Done

- [ ] Documentation has been updated, if applicable.
- [ ] Tests have been added if necessary.
- [ ] Automated tests have passed.
- [ ] Code Quality Tests have passed.
- [ ] Acceptance Criteria in related issue are met.
- [ ] Manual Code Review Approved.
- [ ] If there are field changes, front end output has been thoroughly checked.

### Select Team for PR review

- [x] `CMS Team`
- [ ] `Public websites`
- [ ] `Facilities`
- [ ] `User support`
- [ ] `Accelerated Publishing`

### Is this PR blocked by another PR?

- [ ] `DO NOT MERGE`

### Does this PR need review from a Product Owner

- [ ] `Needs PO review`

### CMS user-facing announcement

Is an announcement needed to let editors know of this change?
- [ ] Yes, and it's written in issue ____ and queued for publication.
  - [ ] Merge and ping the UX writer so they are ready to publish after deployment
- [ ] Yes, but it hasn't yet been written
  - [ ] Don't merge yet -- ping the UX writer to write and queue content
- [ ] No announcement is needed for this code change.
  - [ ] Merge & carry on unburdened by announcements

### CODE REVIEW

**Review by keisterj-oddball — APPROVED** [review]
All testing passed.  Looks good from my end.

File: .tugboat/.env.j2 [hunk 1]
```diff
@@ -16,6 +16,8 @@ CMS_VAGOV_API_URL="{{ CMS_VAGOV_API_URL }}"
 CMS_GOVDELIVERY_ENDPOINT="{{ CMS_GOVDELIVERY_ENDPOINT }}"
 CMS_GOVDELIVERY_USERNAME="{{ CMS_GOVDELIVERY_USERNAME }}"
 CMS_GOVDELIVERY_PASSWORD="{{ CMS_GOVDELIVERY_PASSWORD }}"
+CMS_PDF_SERVCICE_ACCT_KEY="{{ CMS_PDF_SERVICE_ACCT_KEY }}"
```
Comment (timcosgrove) [inline .tugboat/.env.j2]: How hard would it be to correct the spelling here? I don't mean that sarcastically, I mean, how much of a lift?

File: docroot/modules/custom/va_gov_eca/src/Plugin/ECA/Condition/ViewsResultCondition.php [hunk 2]
```diff
@@ -33,7 +33,7 @@ class ViewsResultCondition extends ConditionBase {
    */
   public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition): static {
     $instance = parent::create($container, $configuration, $plugin_id, $plugin_definition);
-    $instance->module_handler = $container->get('module_handler');
+    $instance->moduleHandler = $container->get('module_handler');
```
Comment (timcosgrove) [inline docroot/modules/custom/va_gov_eca/src/Plugin/ECA/Condition/ViewsResultCondition.php:36]: Is this change related to this work? What are we using ECA for here?

File: docroot/modules/custom/va_gov_media/src/Service/PdfDeleteService.php [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+      ->accessCheck(FALSE)
+      ->condition('bundle', 'document')
+      ->execute();
+
+    $total_pdfs = count($pdf_list);
+    // Used for name csv and directory.
+    $timestamp = date('Y-m-d-H:i:s');
+
+    foreach ($pdf_list as $pdf_id) {
+      $is_used = $this->usage->getUsageTotal('media', $pdf_id);
```
Comment (timcosgrove) [inline docroot/modules/custom/va_gov_media/src/Service/PdfDeleteService.php:132]: I assume this is is truthy/falsy, i.e. 0 = false, anything else = true?

File: docroot/modules/custom/va_gov_media/src/Service/PdfDeleteService.php [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+      /** @var \Drupal\media\MediaInterface $pdf */
```
Comment (timcosgrove) [inline docroot/modules/custom/va_gov_media/src/Service/PdfDeleteService.php:133]: Minor, but this block could go before `$is_used` above, so that if the document field check fails you don't bother with the usage total lookup (unless you do something else with `$is_used` later). Small computation savings.

File: docroot/modules/custom/va_gov_media/src/Service/PdfDeleteService.php [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+      $pdf = $media_storage->load($pdf_id);
+      $pdf_name = $pdf->get('name')->value;
+      if (!$pdf->hasField('field_document')) {
+        $logger->warning("PDF {$pdf_id} not deleted because it has no document field.");
+        continue;
+      }
+      if ($is_used === 0 && $pdf->hasField('field_document')) {
+        $section_id = $pdf->get('field_owner')->getValue()[0]['target_id'];
+        if (!$section_id) {
+          $logger->warning("PDF {$pdf_id} has no section id.");
```
Comment (timcosgrove) [inline docroot/modules/custom/va_gov_media/src/Service/PdfDeleteService.php:143]:
```suggestion
          $logger->warning("PDF {$pdf_id} has no section id. Skipping deletion.");
```

File: docroot/modules/custom/va_gov_media/src/Service/PdfDeleteService.php [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+          continue;
+        }
+        $section = $this->entityTypeManager->getStorage('taxonomy_term')->load($section_id);
+        $parents = $this->entityTypeManager->getStorage('taxonomy_term')->loadAllParents($section_id);
+        $parents = array_keys($parents);
+        $section_name = $section ? $section->get('name')->value : 'No Section';
+        // 3 is the Term ID for VHA.
+        if (in_array('8', $parents)) {
```
Comment (timcosgrove) [inline docroot/modules/custom/va_gov_media/src/Service/PdfDeleteService.php:151]: 3 vs 8...which one is correct?

File: docroot/modules/custom/va_gov_media/src/Service/PdfDeleteService.php [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+    $this->dateFormatter = $date_formatter;
+    $this->s3fs = $s3fsfileservice;
+    $this->loggerFactory = $loggerFactory;
+    $this->user = $user;
+  }
+
+  /**
+   * Find and delete PDFs that are not attached to content.
+   */
+  public function pdfDelete(): void {
```
Comment (timcosgrove) [inline docroot/modules/custom/va_gov_media/src/Service/PdfDeleteService.php:112]: This is named `pdfDelete` and the function for deleting a single PDF is `deletePdf`. I might name this `deletePDFs`, or even `deleteVhaPdfs`, since right now this only deletes VHA PDFs and both the Taxonomy ID for that and the error messages are hard-coded into the function.

I don't think it needs to be abstracted right now, but I would recommend getting more explicit in the function name.

File: docroot/modules/custom/va_gov_media/src/Service/PdfDeleteService.php [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+   * @return void
+   *   Return void.
+   *
+   * @throws \Drupal\Core\Entity\EntityStorageException
+   */
+  protected function deletePdf(MediaInterface $pdf_entity, string $timestamp): void {
+    $file = $pdf_entity->field_document->entity;
+    if (!$file instanceof FileInterface) {
+      $pdf_name = $pdf_entity->get('name')->value;
+      $this->loggerFactory->get('va_gov_media')->warning("PDF {$pdf_name} file could not be copied because it has no file entity.");
```
Comment (timcosgrove) [inline docroot/modules/custom/va_gov_media/src/Service/PdfDeleteService.php:197]: If I understand this correctly, we are deleting the PDF _entity_ here because it has no PDF file attached to it. If that's the case, I would amend the logged message as below.

```suggestion
      $this->loggerFactory->get('va_gov_media')->warning("{$pdf_name} does not have a file attached to it. No file was copied. Deleting the file entity.");
```

File: docroot/modules/custom/va_gov_media/src/Service/PdfDeleteService.php [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+      $pdf_entity->delete();
+      return;
+    }
+    $uri = $file->getFileUri();
+    $absolute_path = $this->s3fs->realpath($uri);
+    $path = 's3://' . $timestamp;
```
Comment (timcosgrove) [inline docroot/modules/custom/va_gov_media/src/Service/PdfDeleteService.php:203]: If I understand this, we will have a different timestamp in front of every file we delete. Is that what we want? It will create a directory for every file, which will make browsing for the file difficult. Would it make sense to set a timestamp at the beginning of the `pdfDelete` method that calls this method, so that there is a consistent directory that gathers all deletions?

File: docroot/modules/custom/va_gov_media/src/Service/PdfDeleteService.php [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+
+    // Open a file in write mode ('w')
+    $file = fopen($filename, 'w');
+
+    // Loop through file pointer and a line.
+    foreach ($deleted_pdfs as $fields) {
+      fputcsv($file, $fields);
+    }
+    fclose($file);
+    $destination = 's3://' . $timestamp . '/' . $filename;
```
Comment (timcosgrove) [inline docroot/modules/custom/va_gov_media/src/Service/PdfDeleteService.php:239]: This is one place I would actively recommend against using a timestamp as a directory, because the file will no longer be findable if you do not have access to the logs. The timestamp is already written into the filename. I would either put these in the 'root' of the S3 bucket or else in a named directory `deletion_logs` or `logs` or something like that.

File: phpstan-baseline.neon [hunk 11]
```diff
@@ -496,6 +496,6 @@ parameters:
 			path: tests/scaling_performance.php
 
 		-
-		  message: "#^Call to an undefined method Drupal\\\\Core\\\\\Form\\\\\FormInterface\\:\\:getEntity\\(\\)\\.$#"
```
Comment (timcosgrove) [inline phpstan-baseline.neon:499]: Just curious about this change.

File: docroot/modules/custom/va_gov_eca/src/Plugin/ECA/Condition/ViewsResultCondition.php (hunk 2 above)
Comment (edmund-dunn) [reply docroot/modules/custom/va_gov_eca/src/Plugin/ECA/Condition/ViewsResultCondition.php:36]: this came up in phpstan testing

File: .tugboat/.env.j2 (hunk 1 above)
Comment (edmund-dunn) [reply .tugboat/.env.j2]: fixed

File: docroot/modules/custom/va_gov_media/src/Service/PdfDeleteService.php (hunk 3 above)
Comment (edmund-dunn) [reply docroot/modules/custom/va_gov_media/src/Service/PdfDeleteService.php:132]: It returns a literal count.

File: docroot/modules/custom/va_gov_media/src/Service/PdfDeleteService.php (hunk 6 above)
Comment (edmund-dunn) [reply docroot/modules/custom/va_gov_media/src/Service/PdfDeleteService.php:151]: 8, I need to fix the comment.

File: phpstan-baseline.neon (hunk 11 above)
Comment (edmund-dunn) [reply phpstan-baseline.neon:499]: phpstan was screaming about it.

**Review by timcosgrove — APPROVED** [review]

**Review by keisterj-oddball — APPROVED** [review]
Testing looked good in tugboat.
